File: src/main/java/ua/notion/infrastructure/persistence/contract/UserRepository.java
package ua.notion.infrastructure.persistence.contract;

import java.util.List;
import ua.notion.domain.entity.User;
import ua.notion.infrastructure.persistence.Repository;

public interface UserRepository extends Repository<User, Integer> {

  List<User> findByUsername(String username);

  List<User> findByEmail(String email);

  boolean existsByUsername(String username);

  boolean existsByEmail(String email);
}

File: src/server/src/repositories/contract/user_repository.rs
use axum::async_trait;

use crate::{errors::error::DbResult, models::User, repositories::repository::Repository};

#[async_trait]
pub trait UserRepository: Repository<User, i32> {
    async fn find_by_username(&self, username: &str) -> DbResult<Option<User>>;
    async fn find_by_email(&self, email: &str) -> DbResult<Option<User>>;

    async fn exists_by_username(&self, username: &str) -> DbResult<bool>;
    async fn exists_by_email(&self, email: &str) -> DbResult<bool>;
}
